### PR TITLE
🐛 fix cookies override problems

### LINKED ISSUE #1651: cookie_manager should merge custom cookies instead of overriding them
### Package

cookie_manager

### Version

2.1.0

### Output of `flutter doctor -v`

```shell
$ flutter doctor -v
[✓] Flutter (Channel stable, 3.7.3, on macOS 13.2 22D49 darwin-arm64, locale en-US)
    • Flutter version 3.7.3 on channel stable at /Users/<deleted>/.local/opt/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 9944297138 (6 days ago), 2023-02-08 15:46:04 -0800
    • Engine revision 248290d6d5
    • Dart version 2.19.2
    • DevTools version 2.20.1

[✓] Android toolchain - develop for Android devices (Android SDK version 33.0.2)
    • Android SDK at /Users/<deleted>/.local/opt/android
    • Platform android-33, build-tools 33.0.2
    • ANDROID_HOME = /Users/<deleted>/.local/opt/android
    • Java binary at: /Library/Java/JavaVirtualMachines/temurin-11.jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment Temurin-11.0.18+10 (build 11.0.18+10)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 14.2)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Build 14C18
    • CocoaPods version 1.11.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[!] Android Studio (not installed)
    • Android Studio not found; download from https://developer.android.com/studio/index.html
      (or visit https://flutter.dev/docs/get-started/install/macos#android-setup for detailed instructions).

[✓] IntelliJ IDEA Community Edition (version 2022.1.3)
    • IntelliJ at /Applications/IntelliJ IDEA CE.app
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.75.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension can be installed from:
      🔨 https://marketplace.visualstudio.com/items?itemName=Dart-Code.flutter

[✓] VS Code (version 1.76.0-insider)
    • VS Code at /Users/<deleted>/Applications/Visual Studio Code - Insiders.app/Contents
    • Flutter extension version 3.58.0

[✓] Connected device (2 available)
    • macOS (desktop) • macos  • darwin-arm64   • macOS 13.2 22D49 darwin-arm64
    • Chrome (web)    • chrome • web-javascript • Google Chrome 109.0.5414.119

[✓] HTTP Host Availability
    • All required HTTP hosts are available

! Doctor found issues in 1 category.
```

### Dart Version

2.19.2

### Steps to Reproduce

Cookie manager [overrides the existing `cookie` header](https://github.com/cfug/dio/blob/6f2dcb55c84087d3429e4387e53671c5c38973a3/plugins/cookie_manager/lib/src/cookie_mgr.dart#L29) when it stores cookies for the website. The user provided custom cookie header will be dropped without notice. Is this the expected behavior and what's the reason behind this decision? 

Example code to reproduce the problem:

```dart
final Dio _dio = Dio();
 _dio.interceptors.add(CookieManager(CookieJar()));

// request any website with a 'set-cookie' response header
await _dio.request('https://example.com');

// request the same website again, with a custom cookie
await _dio.request('https://example.com',
  options: Options(
    headers: {'cookie': 'foo=bar'},
  ),
);
```

The second request would only send cookies set by the server from the first request.

### Expected Result

Custom cookies should be send along with server set cookies.

### Actual Result

Only server set cookies are send.

### PR DESCRIPTION
**Pull request by cleong98**
Fix #1651.

### New Pull Request Checklist

- [x] I have read the [Documentation](https://pub.dev/documentation/dio/latest/)
- [x] I have searched for a similar pull request in the [project](https://github.com/cfug/dio/pulls) and found none
- [x] I have updated this branch with the latest `main` branch to avoid conflicts (via merge from master or rebase)
- [x] I have added the required tests to prove the fix/feature I'm adding
- [ ] I have updated the documentation (if necessary)
- [x] I have run the tests without failures

### CODE REVIEW

**Comment by ueman** [comment]
Can you please add a test? Otherwise, this code might get broken in the future once again.

**Comment by cleong98** [comment]
> Thanks for the PR. Can you please add a test? Otherwise, this code might get broken in the future once again.

thanks for review, about testing the cookies problems i don't have found a free or public api for test. i just testing in my local server api.  can you give me some suggestion for testing?

**Comment by ueman** [comment]
Since this is code in an interceptor, you don't actually need to do a real http request.
You can do something like the following pseudo-code:
```dart
class MockRequestInterceptorHandler extends RequestInterceptorHandler {
  // mock required methods
}

final mockRequestInterceptorHandler = MockRequestInterceptorHandler();
final options = RequestOptions(); // set proper values here
final cookieManager = CookieManager();
cookieManager.onRequest(options, mockRequestInterceptorHandler);

// verify here the correct data was passed to the mockRequestInterceptorHandler
```

**Comment by cleong98** [comment]
> Since this is code in an interceptor, you don't actually need to do a real http request. You can do something like the following pseudo-code:
> 
> ```dart
> class MockRequestInterceptorHandler extends RequestInterceptorHandler {
>   // mock required methods
> }
> 
> final mockRequestInterceptorHandler = MockRequestInterceptorHandler();
> final options = RequestOptions(); // set proper values here
> final cookieManager = CookieManager();
> cookieManager.onRequest(options, mockRequestInterceptorHandler);
> 
> // verify here the correct data was passed to the mockRequestInterceptorHandler
> ```

thanks for suggest i add a test now

File: plugins/cookie_manager/lib/src/cookie_mgr.dart [hunk 1]
```diff
@@ -24,9 +24,19 @@ class CookieManager extends Interceptor {
...
+            .map((cookie) => Cookie.fromSetCookieValue(cookie))
+            .toList();
+        options.headers[HttpHeaders.cookieHeader] =
+            getCookies(<Cookie>[...cookies, ...newCookies]);
+      } else {
+        final cookie = getCookies(cookies);
+        if (cookie.isNotEmpty) {
+          options.headers[HttpHeaders.cookieHeader] = cookie;
+        }
       }
```
Comment (AlexV525) [inline plugins/cookie_manager/lib/src/cookie_mgr.dart:40]: The whole flow seems can be simplified by:

```dart
options.headers[HttpHeaders.cookieHeader] = getCookies([
  ...cookies,
  ...?options.headers[HttpHeaders.cookieHeader]
      ?.split(';')
      .map((c) => Cookie.fromSetCookieValue(c)),
]);
```

Comment (cleong98) [reply plugins/cookie_manager/lib/src/cookie_mgr.dart:40]: > The whole flow seems can be simplified by:
> 
> ```dart
> options.headers[HttpHeaders.cookieHeader] = getCookies([
>   ...cookies,
>   ...?options.headers[HttpHeaders.cookieHeader]
>       ?.split(';')
>       .map((c) => Cookie.fromSetCookieValue(c)),
> ]);
> ```

looks good

File: plugins/cookie_manager/test/cookies_test.dart [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        headers: Headers.fromMap(
+          {HttpHeaders.setCookieHeader: mockFirstRequestCookies},
+        ));
+    cookieManager.onResponse(mockResponse, mockResponseInterceptorHandler);
+    final options = RequestOptions(baseUrl: exampleUrl, headers: {
+      HttpHeaders.cookieHeader: mockSecondRequestCookies,
+    });
+
+    cookieManager.onRequest(options, mockRequestInterceptorHandler);
+  });
```
Comment (ueman) [inline plugins/cookie_manager/test/cookies_test.dart:53]:
```suggestion
class MockRequestInterceptorHandler extends RequestInterceptorHandler {
  final String expectResult;
  
  MockRequestInterceptorHandler(this.expectResult);
  
  @override
  void next(RequestOptions requestOptions) {
    final c = requestOptions.headers[HttpHeaders.cookieHeader];
    expect(c == expectResult, true);
  }
}

class MockResponseInterceptorHandler extends ResponseInterceptorHandler {}

void main() {
  test('testing merge cookies', () async {
    const List<String> mockFirstRequestCookies = [
      "foo=bar; Path=/",
      "a=c; Path=/"
    ];
    const exampleUrl = 'https://example.com';
    const String mockSecondRequestCookies = 'd=e;e=f';

    final expectResult = mockFirstRequestCookies
            .map((mock) => mock.replaceAll(' Path=/', ' '))
            .join() +
        mockSecondRequestCookies.split(';').join('; ');

    final cookieJar = CookieJar();
    final cookieManager = CookieManager(cookieJar);
    final mockRequestInterceptorHandler =
        MockRequestInterceptorHandler(expectResult);
    final mockResponseInterceptorHandler = MockResponseInterceptorHandler();
    final firstRequestOptions = RequestOptions(baseUrl: exampleUrl);

    cookieManager.onRequest(options, mockRequestInterceptorHandler);
  });
```

**Review by ueman — COMMENTED** [review]
Thanks for adding the test. While the test can be further improved, it does test the new feature, so it's an approvement compared to before. I'm tempted to approve, what do the others think?

File: plugins/cookie_manager/test/cookies_test.dart [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+    final mockResponse = Response(
+        requestOptions: firstRequestOptions,
+        headers: Headers.fromMap(
+          {HttpHeaders.setCookieHeader: mockFirstRequestCookies},
+        ));
+    cookieManager.onResponse(mockResponse, mockResponseInterceptorHandler);
+    final options = RequestOptions(baseUrl: exampleUrl, headers: {
+      HttpHeaders.cookieHeader: mockSecondRequestCookies,
+    });
```
[the remark this answers is not in the document]
Comment (cleong98) [reply plugins/cookie_manager/test/cookies_test.dart:50]: > Is this actually needed?

hihi thanks for review, actually these code purpose is to mock response with cookies and save it to manager. if we don't have these code there are another way to mock exist cookies in second request?

[the remark this answers is not in the document]
Comment (ueman) [reply plugins/cookie_manager/test/cookies_test.dart:50]: Thanks, I missed that.

**Review by ueman — APPROVED** [review]

**Comment by AlexV525** [comment]
You'll need to update the CHANGELOG.md too.

**Comment by cleong98** [comment]
@AlexV525 update CHANGELOG already

File: plugins/cookie_manager/CHANGELOG.md [hunk 4]
```diff
@@ -1,5 +1,9 @@
 # CHANGELOG
 
+# 2.1.1
```
Comment (AlexV525) [inline plugins/cookie_manager/CHANGELOG.md:3]: And the pubspec. :)

Comment (cleong98) [reply plugins/cookie_manager/CHANGELOG.md:3]: > And the pubspec. :)

thanks for remind, pubspec update already.

**Review by AlexV525 — APPROVED** [review]

**Pull request merged**